OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs

```diff
@@ -93,17 +93,22 @@ public void Attack(Actor targetActor, bool allowMove = true, bool forceAttack =
 			if (!target.IsValidFor(Self))
 				Log.Write("lua", "{1} is an invalid target for {0}!", Self, targetActor);
 
-			if (!targetActor.Info.HasTraitInfo<FrozenUnderFogInfo>() && !targetActor.CanBeViewedByPlayer(Self.Owner))
-				Log.Write("lua", "{1} is not revealed for player {0}!", Self.Owner, targetActor);
+			if (!IsActorVisible(targetActor))
+				Log.Write("lua", "{2} is not revealed for player {1}'s {0}!", Self, Self.Owner, targetActor);
 
 			foreach (var attack in attackBases)
 				attack.AttackTarget(target, true, allowMove, forceAttack);
 		}
 
-		[Desc("Checks if the targeted actor is a valid target for this actor.")]
+		[Desc("Checks if the targeted actor is a valid, visible target for this actor.")]
 		public bool CanTarget(Actor targetActor)
 		{
-			return Target.FromActor(targetActor).IsValidFor(Self);
+			return Target.FromActor(targetActor).IsValidFor(Self) && IsActorVisible(targetActor);
+		}
+
+		bool IsActorVisible(Actor targetActor)
+		{
+			return targetActor.Info.HasTraitInfo<FrozenUnderFogInfo>() || targetActor.CanBeViewedByPlayer(Self.Owner);
```

Comment:
For example, the check in `CanTarget` above can hopefully be replaced by `Target.FromActor(targetActor).Recalculate(Self.Owner).IsValidFor(Self)`.

Reply:
Okay, I'll continue to look into this part. I did once see a case where a yak's recalculated target in FlyAttack became invalid, so making more direct use of that system here definitely sounds good. 

In the meantime should I also split out the lua changes to another PR? Since the CallFunc loop there is fatal.

Reply:
Best to keep this as the PR for lua changes. I'll file a separate PR for the visibility fixes (see #16010 / #10223) in a few days time, and then we can rebase and retest this one on top.

Reply:
Okay, I pulled out changes to CombatProperties. I also pulled the Reservable changes out into https://github.com/OpenRA/OpenRA/pull/16055. 

The only thing left in this PR is the lua changes. This alone won't provide reliable AI yaks, but at least the level won't hang in CallFunc.

I realized that since I later changed the lua script from using Attack to AttackMove, this CombatProperties visibility check is not necessary anymore. But fixing the asymmetry between CanTarget and Attack, and updating it with the new system might still be nice.